### DISCUSSION TITLE
model in ddp

### DISCUSSION THREAD

**Discussion by valiantljk** [allenai/allennlp#5238] [Q&A] [answered]
Hi,
I noticed that in distributed training, there will be two models, one is the original model, and another is the wrapped model. For example, https://github.com/allenai/allennlp/blob/c5bff8ba0d835eb03931f10f4f427ffe936cf796/allennlp/training/gradient_descent_trainer.py#L351

I have a general question (not specific to allennlp), in this case, can we make sure the two versions of model share the same parameters? 
In runtime, is this some python magic, that they both refer to the same object but each has some customization? I'm a bit confused here.

**Comment by epwalsh** [comment, marked as answer]
Yes, they share the same parameters since the `_pytorch_model` is just a thin wrapper around the `model`. In other words, `_pytorch_model` just holds a (mutable) reference to `model`, it doesn't clone the parameters.

**Reply by valiantljk** [reply]
Thanks for the quick response!

**Reply by valiantljk** [reply]
btw, when the _pytorch_model is copied onto cuda device, does the self.original_model still hold the reference to it? and can access it when computing get_metrics.

**Reply by epwalsh** [reply]
`model` is put on the correct device before the `_pytorch_model` is created, so the `_pytorch_model` never has to be copied onto the target device.

**Reply by valiantljk** [reply]
That makes sense. In my case, I'm using Ray to take care the distribute training setup, in which I'll let Ray to auto-copy the model on selected GPU. I'm wondering in this case, what will happen to the original model?

**Reply by valiantljk** [reply]
Also, why those customized functions, e.g., get_metrics will be lost in the wrapped ddp model?

**Reply by epwalsh** [reply]
I'm not familiar with Ray so I'm not sure.

> Also, why those customized functions, e.g., get_metrics will be lost in the wrapped ddp model?

The `_pytorch_model` is just an instance of `torch.nn.parallel.DistributedDataParallel`. That class doesn't have a `.get_metrics()` method. We could call `_pytorch_model.module.get_metrics()` though, since `_pytorch_model.module` is just a reference to `model`.

**Reply by valiantljk** [reply]
Cool, Thanks @epwalsh. So can I use `self._pytorch_model.module.get_regularization_penalty()` to replace `self.model.get_regularization_penalty()`
as in https://github.com/allenai/allennlp/blob/c5bff8ba0d835eb03931f10f4f427ffe936cf796/allennlp/training/gradient_descent_trainer.py#L356, 
If so, why do we need to keep two reference?

**Reply by epwalsh** [reply]
Yeup, you can do that. It's not necessary to keep two references, but it doesn't hurt either and makes things slightly simpler I think since `_pytorch_model.module` is only valid in the distributed context. In the non-distributed context, `_pytorch_model` is just `model`.

**Reply by valiantljk** [reply]
Got it! Very appreciate!
